### ISSUE TITLE
render_to_string changes content-type for subsequent render

### ISSUE THREAD

**Issue by luckyruby** [rails/rails#14173]
In Rails 3.2.16:

``` ruby
  def locations
    @locations = Location.all
    stream = render_to_string(template: "locations/index", formats: [:xml])
    render nothing: true
  end
```

I would expect a blank page when I hit this controller action from my browser but I get an XML parsing error in Firefox instead. It seems that render_to_string is changing the content-type response header. Is that expected behavior?

**Comment by manufaktor** [comment]
I can reproduce the failing test on Rails 4.2.1 (Ruby 2.1.5) and Rails master (Ruby 2.2.1).

**Comment by senny** [comment]
@manufaktor thank you for confirming. I removed the `stale` label and added `pinned`.

**Comment by lulzmachine** [comment]
Happening here too, in 4.2. I'm rendering some jbuilder json like so: @json = render_to_string template: "browse/_json", layout: false

And it causes the whole page to render with content-type application/json.

Please let me know if you need more input.

**Comment by manufaktor** [comment]
So I tried to debug this, but this was my first time digging into rails internals and I got a bit lost. 

What I found is that the `PartialRenderer` does a conditional assignment of `rendered_format=` onto the `LookupContext`, which is shared between `render_to_string` and `render`. I guess removing the conditional assignment in `actionview/lib/action_view/renderer/partial_renderer.rb:299` would fix the test I've provided.

**Comment by manufaktor** [comment]
I guess it would also be possible to set `lookup_context.rendered_format = nil` when `render_to_string` is done. It would then be freshly assigned again when calling `render.

**Comment by steve32** [comment]
In 4.1.8: `render_to_string` is causing the response to my xhr request to have the following header: `Content-Type:text/html`.

Removing the call to `render_to_string` and the header normalizes to: `Content-Type:text/javascript`.

**Comment by repinel** [comment]
FWIW, minimal example of the issue using Rails 5 `master`: https://gist.github.com/repinel/92810d0d47ec23543016

**Comment by repinel** [comment]
Since `render_to_string` reuses most of the `render_template` features, it seems to me that the `rendered_format` is being set [here](https://github.com/rails/rails/blob/2bf05de5ab5ea63eccd819e57200a80c2e6ee8e6/actionview/lib/action_view/renderer/template_renderer.rb#L12).

@senny Could we provide some option around [here](https://github.com/rails/rails/blob/2bf05de5ab5ea63eccd819e57200a80c2e6ee8e6/actionpack/lib/abstract_controller/rendering.rb#L43) indicating that the result will be rendered as `string` and there is no need to set the `rendered_format`? Any other suggestion?

**Comment by jcoleman** [comment]
This is also linked to #14693

**Comment by maclover7** [comment]
Been a long time coming, but opened up PR #24665 to solve this. :)

**Comment by slemrmartin** [comment]
> lookup_context.rendered_format = nil when render_to_string is done 

It works for Rails 4.2.6

**Comment by raymondji** [comment]
Any follow up on how this is progressing?

**Comment by ramaboo** [comment]
I am having this issue as well! Fix would be awesome!

**Comment by rewolf** [comment]
Also still seeing this on 5.1.0.  If I use `render_to_string` somewhere in an action then the overall format is thought to be plaintext when the following render occurs for the action

**Comment by kevmimcc** [comment]
Any work around for this? I'm still getting this with:
Rails 5.0.2
Ruby 2.4.0p0

**Comment by GermanDZ** [comment]
I've used this trick in the past to get rid of this problem:

```
def some_action
  original_formats = formats.dup

  # This is the render under a new format
  info_as_json = JSON.parse(render_to_string(formats: [:json], layout: false))

  # This is a hack to force to the controller to render the right format.
  # The previous render_to_string seems to be changing the format for the whole request
  render_to_string(template: "module_name/controller/action", formats: original_formats, layout: false)

  .
  .
  the rest of your action (probably nothing, the action will render the default template, in this case `some_action`
end
```

**Comment by mrj** [comment]
I work around this bug by explicitly setting the ```content_type``` for the action template render at the top of any controller action that calls ```render_to_string```.

For example, if you're rendering a JavaScript action template, add ```response.content_type = 'text/javascript'``` to the top of the action to prevent ```render_to_string``` of HTML templates during the action from setting the final ```content_type``` to ```text/html```, which stops the JavaScript from executing.

**Comment by GermanDZ** [comment]
Good to know, but in my case I'm responding with text or json (another kind of text :) ) with the same code, the view is in charge of the formatting. So, I can't "force" a single format in the action method.

**Comment by mrj** [comment]
If you have multiple render templates you can add
```
respond_to do |format|
  format.html
  format.json
end
```
to the bottom of the action, since this also sets the correct ```content_type``` from the request.

I've verified that adding
```
respond_to { |format| format.js }
```
works in my case.

**Comment by jplaut** [comment]
Still getting this with Rails 5.1.4

**Comment by laaksomavrick** [comment]
@mrj You're a life saver, thank you!

**Comment by vizcay** [comment]
Still getting this with Rails 5.1.2.

Basic example:
```
  def edit
    @model = Model.find(params[:id])
    @data = render_to_string('data.json')
  end
```

I had to add ```render 'edit', content_type: 'text/html'``` at the end to force a HTML Content-Type response.

**Comment by hahmed** [comment]
I'm wondering if this is still an issue?

I tested this and it seems to work for me when I use inline or template. Let me know if there is something wrong with my repro script...

```
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"
  git_source(:github) { |repo| "https://github.com/#{repo}.git" }
  gem "rails", "~>6.0"
end

require "rack/test"
require "action_controller/railtie"
require "minitest/autorun"

class TestApp < Rails::Application
  config.root = __dir__
  config.hosts << "example.org"
  config.session_store :cookie_store, key: "cookie_store_key"
  secrets.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    get "/" => "test#index"
  end
end

class TestController < ActionController::Base
  include Rails.application.routes.url_helpers
  before_action :prepend_view_paths

   def prepend_view_paths
     prepend_view_path "#{Rails.root}"
   end

  def index
    template_string = render_to_string(partial: "index", formats: [:xml], layout: false)
    render plain: "Home"
  end

  def index2
    inline_template = render_to_string(inline: <<-XML, formats: [:xml])
      <div>WHAT</div>
    XML
    render plain: "Home"
  end
end

class BugTest < Minitest::Test
  include Rack::Test::Methods

  def test_returns_success
    get "/"
    assert_equal "Home", last_response.body
    assert_equal "text/plain; charset=utf-8", last_response.content_type
  end

  private
    def app
      Rails.application
    end
end
```

The output I'm getting for `last_response.inspect` is:

```
#<Rack::MockResponse:0x00007fc4f4ba0cc8 @original_headers={"Content-Type"=>"text/plain; charset=utf-8", "ETag"=>"W/\"3a78695388b38b5cceefaf6796b01378\"", "Cache-Control"=>"max-age=0, private, must-revalidate", "X-Request-Id"=>"c5f4e063-7659-49ca-ad04-64711db0e16c", "X-Runtime"=>"0.114918", "Content-Length"=>"4"}, @errors="", @cookies={}, @status=200, @headers={"Content-Type"=>"text/plain; charset=utf-8", "ETag"=>"W/\"3a78695388b38b5cceefaf6796b01378\"", "Cache-Control"=>"max-age=0, private, must-revalidate", "X-Request-Id"=>"c5f4e063-7659-49ca-ad04-64711db0e16c", "X-Runtime"=>"0.114918", "Content-Length"=>"4"}, @writer=#<Method: Rack::MockResponse(Rack::Response::Helpers)#append(chunk) /Users/haroon/.gem/ruby/3.0.2/gems/rack-2.2.3/lib/rack/response.rb:287>, @block=nil, @body=["Home"], @buffered=true, @length=4>
```

**Comment by joshuapinter** [comment]
For what it's worth, I'm seeing this fixed on Rails 7.0.4. We've now removed it from our codebase after upgrading and things seem to be working fine. Thanks for the fix. :+1:
